# Superstore Dataset — Exploration & Cleaning with Pandas

**Author:** Vabhravi Pandey

**Objective:** Practice core Pandas workflows — loading, exploring, and cleaning a real-world
messy dataset end to end.

**Dataset:** [Superstore Dataset Final](https://www.kaggle.com/datasets/vivek468/superstore-dataset-final) (Kaggle)

**Steps covered:**
1. Load CSV into a DataFrame
2. Explore data (head/tail, shape, columns, dtypes)
3. Handle missing values
4. Filter rows / select columns
5. Remove duplicates
6. Create a derived column (`total_amount`)
7. Save the cleaned dataset


## 1. Load the dataset

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv("Superstore_raw.csv", encoding="latin1")
print(f"Rows: {df.shape[0]:,} | Columns: {df.shape[1]}")
df.head()

Rows: 10,800 | Columns: 21


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2017-152156,11/8/2017,11/11/2017,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420.0,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2.0,0.00,41.9136
1,2,CA-2017-152156,11/8/2017,11/11/2017,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420.0,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3.0,0.00,219.5820
2,3,CA-2017-138688,6/12/2017,6/16/2017,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036.0,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2.0,0.00,6.8714
3,4,US-2016-108966,10/11/2016,10/18/2016,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311.0,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5.0,0.45,-383.0310
4,5,US-2016-108966,10/11/2016,10/18/2016,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311.0,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2.0,0.20,2.5164


## 2. Explore the data
Basic structural checks before touching anything.

In [2]:
print("Shape:", df.shape)
print("\nColumns:", list(df.columns))
print("\nData types:\n", df.dtypes)

Shape: (10800, 21)

Columns: ['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode', 'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State', 'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category', 'Product Name', 'Sales', 'Quantity', 'Discount', 'Profit']

Data types:
 Row ID               str
Order ID             str
Order Date           str
Ship Date            str
Ship Mode            str
Customer ID          str
Customer Name        str
Segment              str
Country              str
City                 str
State                str
Postal Code      float64
Region               str
Product ID           str
Category             str
Sub-Category         str
Product Name         str
Sales            float64
Quantity         float64
Discount         float64
Profit           float64
dtype: object


In [3]:
df.tail()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
10795,Yes,US-2018-147886,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
10796,Yes,US-2018-147998,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
10797,Yes,US-2018-151127,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
10798,Yes,US-2018-155999,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
10799,Yes,US-2018-155999,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10800 entries, 0 to 10799
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         10800 non-null  str    
 1   Order ID       10800 non-null  str    
 2   Order Date     9994 non-null   str    
 3   Ship Date      9994 non-null   str    
 4   Ship Mode      9994 non-null   str    
 5   Customer ID    9994 non-null   str    
 6   Customer Name  9994 non-null   str    
 7   Segment        9994 non-null   str    
 8   Country        9994 non-null   str    
 9   City           9994 non-null   str    
 10  State          9994 non-null   str    
 11  Postal Code    9983 non-null   float64
 12  Region         9994 non-null   str    
 13  Product ID     9994 non-null   str    
 14  Category       9994 non-null   str    
 15  Sub-Category   9994 non-null   str    
 16  Product Name   9994 non-null   str    
 17  Sales          9994 non-null   float64
 18  Quantity       99

## 3. Handle missing values

### 3a. Identify missing values

In [5]:
missing = df.isnull().sum()
missing_pct = (missing / len(df) * 100).round(2)
pd.DataFrame({"missing_count": missing, "missing_pct": missing_pct}).query("missing_count > 0")

,missing_count,missing_pct
Order Date,806,7.46
Ship Date,806,7.46
Ship Mode,806,7.46
Customer ID,806,7.46
Customer Name,806,7.46
Segment,806,7.46
Country,806,7.46
City,806,7.46
State,806,7.46
Postal Code,817,7.56


Nearly every column has ~806 missing values concentrated at the **bottom of the file**. Let's look at those rows directly.

In [6]:
df[df["Order Date"].isnull()][["Row ID", "Order ID", "Order Date"]].head(10)

,Row ID,Order ID,Order Date
9994,Person,Region,NaN
9995,Anna Andreadi,West,NaN
9996,Chuck Magee,East,NaN
9997,Kelly Williams,Central,NaN
9998,Cassandra Brandow,South,NaN
9999,Returned,Order ID,NaN
10000,Yes,CA-2015-100762,NaN
10001,Yes,CA-2015-100762,NaN
10002,Yes,CA-2015-100762,NaN
10003,Yes,CA-2015-100762,NaN


**What I found:** rows from index 9994 onward aren't real orders at all — they're leftover rows from *other* sheets
in the original Excel workbook (a "People" sheet and a "Returns" sheet) that got concatenated into this CSV by
mistake. Their `Row ID` values are things like `"Person"`, `"Anna Andreadi"`, `"Returned"`, `"Yes"` — not numeric IDs.

Since these rows are structurally invalid for this table (not just "missing a value here and there"), I decided the
right fix was to **drop them entirely** rather than try to fill them in.

In [7]:
# Row ID should always be numeric for genuine order rows — use that to isolate the real data
df["Row ID"] = pd.to_numeric(df["Row ID"], errors="coerce")
df = df[df["Row ID"].notnull()].copy()
df["Row ID"] = df["Row ID"].astype(int)
print(f"Shape after dropping non-order rows: {df.shape}")

Shape after dropping non-order rows: (9994, 21)


In [8]:
# Re-check missing values on the cleaned rows
df.isnull().sum()[df.isnull().sum() > 0]

Postal Code    11
dtype: int64

### 3b. Fill the remaining missing values
Only `Postal Code` has real gaps now. Since postal code correlates with City/State, I filled each missing
value with the most common postal code for that City (falling back to 0 if a city has no other rows).

In [9]:
city_postal = df.groupby("City")["Postal Code"].agg(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)
df["Postal Code"] = df.apply(
    lambda row: city_postal.get(row["City"]) if pd.isnull(row["Postal Code"]) else row["Postal Code"],
    axis=1
)
df["Postal Code"] = df["Postal Code"].fillna(0).astype(int)
print("Remaining missing values:", df.isnull().sum().sum())

Remaining missing values: 0


### 3c. Fix data types
`Order Date` / `Ship Date` should be real dates, not strings.

In [10]:
df["Order Date"] = pd.to_datetime(df["Order Date"], format="%m/%d/%Y")
df["Ship Date"] = pd.to_datetime(df["Ship Date"], format="%m/%d/%Y")
df.dtypes[["Order Date", "Ship Date", "Postal Code"]]

Order Date     datetime64[us]
Ship Date      datetime64[us]
Postal Code             int64
dtype: object

## 4. Basic operations — filter rows & select columns

In [11]:
# Filter: high-value Technology orders shipped Same Day
tech_same_day = df[(df["Category"] == "Technology") & (df["Ship Mode"] == "Same Day")]
print(f"Technology + Same Day orders: {len(tech_same_day)}")
tech_same_day[["Order ID", "Product Name", "Sales", "Quantity"]].head()

Technology + Same Day orders: 98


,Order ID,Product Name,Sales,Quantity
683,US-2018-168116,Cubify CubeX 3D Printer Triple Head Print,7999.98,4.0
792,CA-2017-105256,Cisco Unified IP Phone 7945G VoIP phone,1363.96,5.0
814,CA-2018-143686,Razer Tiamat Over Ear 7.1 Surround Sound PC Ga...,1399.93,7.0
1001,CA-2016-124891,Plantronics CS510 - Over-the-Head monaural Wir...,2309.65,7.0
1149,CA-2016-112452,Shocksock Galaxy S4 Armband,10.95,1.0


In [12]:
# Select a subset of columns for a lightweight summary view
summary_cols = df[["Order ID", "Order Date", "Customer Name", "Category", "Sales", "Quantity", "Profit"]]
summary_cols.head()

,Order ID,Order Date,Customer Name,Category,Sales,Quantity,Profit
0,CA-2017-152156,2017-11-08,Claire Gute,Furniture,261.9600,2.0,41.9136
1,CA-2017-152156,2017-11-08,Claire Gute,Furniture,731.9400,3.0,219.5820
2,CA-2017-138688,2017-06-12,Darrin Van Huff,Office Supplies,14.6200,2.0,6.8714
3,US-2016-108966,2016-10-11,Sean O'Donnell,Furniture,957.5775,5.0,-383.0310
4,US-2016-108966,2016-10-11,Sean O'Donnell,Office Supplies,22.3680,2.0,2.5164


## 5. Remove duplicates

In [13]:
dupe_count = df.duplicated().sum()
print(f"Exact duplicate rows: {dupe_count}")
df = df.drop_duplicates()
print(f"Shape after de-duplication: {df.shape}")

Exact duplicate rows: 0
Shape after de-duplication: (9994, 21)


## 6. Create a derived column — `total_amount`

The raw dataset stores `Sales` as the **line total** for each order row, not a per-unit price, so there's no
`price` column to multiply directly. To satisfy `total_amount = price * quantity` in a way that's honest about
the data, we first back out an implied **unit price**, then recompute `total_amount` from it (which should land
very close to the original `Sales` figure — a nice sanity check).

In [14]:
df["unit_price"] = (df["Sales"] / df["Quantity"]).round(2)
df["total_amount"] = (df["unit_price"] * df["Quantity"]).round(2)

df[["Sales", "Quantity", "unit_price", "total_amount"]].head()

,Sales,Quantity,unit_price,total_amount
0,261.9600,2.0,130.98,261.96
1,731.9400,3.0,243.98,731.94
2,14.6200,2.0,7.31,14.62
3,957.5775,5.0,191.52,957.60
4,22.3680,2.0,11.18,22.36


In [15]:
# Sanity check: total_amount should reconcile with Sales (within rounding)
diff = (df["total_amount"] - df["Sales"]).abs()
print("Max discrepancy between total_amount and original Sales:", diff.max())

Max discrepancy between total_amount and original Sales: 0.05600000000004002


## 7. Save the cleaned dataset

In [16]:
df.to_csv("Superstore_cleaned.csv", index=False)
print(f"Saved Superstore_cleaned.csv — {df.shape[0]:,} rows, {df.shape[1]} columns")

Saved Superstore_cleaned.csv — 9,994 rows, 23 columns


## Summary

- **Started with:** 10,800 rows / 21 columns (raw file included corrupted trailing rows from other sheets)
- **Dropped:** 806 non-order rows that were leftover data from a "People"/"Returns" sheet, wrongly concatenated
  into the Orders table (identified via non-numeric `Row ID`)
- **Filled:** 11 missing `Postal Code` values using the most common postal code per City
- **Fixed dtypes:** `Order Date` / `Ship Date` converted from strings to proper datetimes; `Postal Code` cast to int
- **Duplicates removed:** 0 exact duplicate rows found among genuine order records
- **Derived column added:** `unit_price` and `total_amount` (= `unit_price * Quantity`), verified to reconcile with
  the original `Sales` column
- **Final cleaned dataset:** 9,994 rows / 23 columns, saved to `Superstore_cleaned.csv`
